File: src/llm_cal/core/evaluator.py
"""Evaluator — the single orchestration layer.

v0.1 partial implementation: composes model_source + detector + weight_analyzer
+ reconciler + kv_cache + engine_compat + hardware. Fleet planner and command
generator land in Week 5 remainder.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from llm_cal.architecture.detector import detect
from llm_cal.architecture.formulas.kv_cache import compute_kv_cache_bytes
from llm_cal.architecture.formulas.weight import estimate_total_params
from llm_cal.architecture.profile import ArchitectureProfile
from llm_cal.command_generator.sglang import generate_sglang_command
from llm_cal.command_generator.vllm import generate_vllm_command
from llm_cal.core.cache import ArtifactCache, CacheKey
from llm_cal.engine_compat.loader import EngineCompatEntry, find_match
from llm_cal.fleet.planner import FleetRecommendation, plan
from llm_cal.hardware.loader import GPUSpec, UnknownGPUError, lookup
from llm_cal.model_source.base import ModelArtifact, ModelSource
from llm_cal.model_source.huggingface import HuggingFaceSource
from llm_cal.output.labels import AnnotatedValue
from llm_cal.performance.compute import (
    DEFAULT_DECODE_BW_UTILIZATION,
    DEFAULT_PREFILL_UTILIZATION,
    DecodeEstimate,
    PrefillEstimate,
    estimate_decode,
    estimate_prefill,
)
from llm_cal.performance.concurrency import ConcurrencyAnalysis
from llm_cal.performance.concurrency import analyze as analyze_concurrency
from llm_cal.weight_analyzer import WeightReport, analyze
from llm_cal.weight_analyzer.fingerprint import (
    QuantFingerprint,
    from_config,
    from_safetensors_dtypes,
)
from llm_cal.weight_analyzer.reconciler import ReconciliationReport, reconcile
from llm_cal.weight_analyzer.safetensors_reader import (
    fetch_tensor_dtypes,
    pick_sample_shard,
)

_KV_REFERENCE_CTX = 131_072  # matches fleet.planner's _REFERENCE_CTX_TOKENS


@dataclass(frozen=True)
class EvaluationReport:
    """Everything the evaluator produces for one model."""

    model_id: str
    source: str
    commit_sha: str | None
    gpu: str
    gpu_spec: GPUSpec | None
    gpu_error: str | None  # message if gpu wasn't found
    engine: str
    profile: ArchitectureProfile
    weight: WeightReport
    total_params_estimate: AnnotatedValue[int]
    reconciliation: ReconciliationReport
    kv_cache_by_context: dict[int, AnnotatedValue[int]] = field(default_factory=dict)
    engine_match: EngineCompatEntry | None = None
    fleet: FleetRecommendation | None = None
    generated_command: str | None = None
    # Performance analysis — filled when user passes SLA args (or defaults).
    prefill: PrefillEstimate | None = None
    decode: DecodeEstimate | None = None
    concurrency: ConcurrencyAnalysis | None = None
    perf_input_tokens: int | None = None
    perf_output_tokens: int | None = None
    perf_target_tokens_per_sec: float | None = None


class Evaluator:
    """Orchestrates: model_source -> detect -> analyze -> reconcile -> KV cache
    -> engine compat -> hardware lookup.

    Fleet planning and command generation are remaining Week 5 additions.
    """

    def __init__(
        self,
        source: ModelSource | None = None,
        cache: ArtifactCache | None = None,
    ) -> None:
        self._source = source or HuggingFaceSource()
        self._cache = cache or ArtifactCache()

    def evaluate(
        self,
        model_id: str,
        gpu: str,
        engine: str,
        gpu_count: int | None = None,
        context_length: int | None = None,
        refresh: bool = False,
        input_tokens: int | None = None,
        output_tokens: int | None = None,
        target_tokens_per_sec: float | None = None,
        prefill_utilization: float = DEFAULT_PREFILL_UTILIZATION,
        decode_bw_utilization: float = DEFAULT_DECODE_BW_UTILIZATION,
        concurrency_degradation: float = 1.0,
    ) -> EvaluationReport:
        artifact = self._fetch(model_id, refresh=refresh)
        profile = detect(artifact.config)

        total_params_est = estimate_total_params(profile)
        total_params = total_params_est.value

        observed_bytes_for_fp = sum(
            (s.size or 0) for s in artifact.siblings if s.filename.endswith(".safetensors")
        )
        fingerprint = self._resolve_quant_fingerprint(
            artifact,
            observed_bytes=observed_bytes_for_fp,
            total_params=total_params if total_params > 0 else 0,
        )
        weight = analyze(
            artifact.siblings,
            total_params=total_params if total_params > 0 else None,
            fingerprint=fingerprint,
        )
        reconciliation = reconcile(
            weight.total_bytes.value,
            total_params or 1,
            fingerprint=fingerprint,
        )

        contexts_to_report = self._select_context_lengths(profile, context_length)
        kv_by_ctx = {
            ctx: compute_kv_cache_bytes(profile, seq_len=ctx, dtype_bytes=2)
            for ctx in contexts_to_report
        }

        # Engine compatibility — match by model_type alone (v0.1). Version
        # filtering can be added via a future --engine-version flag.
        engine_match = find_match(engine=engine, model_type=profile.model_type)

        # Hardware lookup — never raises out to CLI, we embed the error message
        # so the user sees a partial report instead of aborting.
        gpu_spec: GPUSpec | None = None
        gpu_error: str | None = None
        try:
            gpu_spec = lookup(gpu)
        except UnknownGPUError as e:
            gpu_error = str(e)

        # Fleet planning — only if we have a known GPU. The planner's reference
        # context is 128K; derive KV bytes there (computing fresh in case the
        # user chose a non-overlapping context_length override).
        fleet: FleetRecommendation | None = None
        generated_command: str | None = None
        if gpu_spec is not None and weight.total_bytes.value > 0:
            kv_ref = compute_kv_cache_bytes(profile, _KV_REFERENCE_CTX, dtype_bytes=2)
            kv_by_context_bytes = {ctx: av.value for ctx, av in kv_by_ctx.items() if av.value > 0}
            fleet = plan(
                profile=profile,
                weight_bytes=weight.total_bytes.value,
                kv_bytes_per_request_at_ref=max(1, kv_ref.value),
                gpu=gpu_spec,
                forced_gpu_count=gpu_count,
                kv_bytes_by_context=kv_by_context_bytes,
            )
            # Pick the gpu_count to emit the command for: user's forced value,
            # else the best_tier's recommendation.
            chosen_count = gpu_count or next(
                (o.gpu_count for o in fleet.options if o.tier == fleet.best_tier),
                fleet.options[0].gpu_count,
            )
            generated_command = _generate_command(
                engine=engine,
                model_id=model_id,
                profile=profile,
                tp=chosen_count,
                entry=engine_match,
                max_model_len=context_length,
            )

        # Performance analysis — runs whenever we have hardware + fleet.
        prefill_est: PrefillEstimate | None = None
        decode_est: DecodeEstimate | None = None
        concurrency_est: ConcurrencyAnalysis | None = None
        if gpu_spec is not None and fleet is not None and total_params > 0:
            # Pick the fleet tier we're analyzing (user's forced count or best tier).
            chosen = gpu_count or next(
                (o.gpu_count for o in fleet.options if o.tier == fleet.best_tier),
                fleet.options[0].gpu_count,
            )
            # Resolve performance defaults when user didn't specify.
            eff_input = input_tokens or 2000
            eff_target = target_tokens_per_sec or 30.0

            prefill_est = estimate_prefill(
                profile=profile,
                total_params=total_params,
                gpu=gpu_spec,
                num_gpus=chosen,
                input_tokens=eff_input,
                utilization=prefill_utilization,
            )
            # MoE active ratio: active/total = (shared + experts_per_tok) / (shared + routed)
            moe_active_ratio: float | None = None
            if profile.moe is not None:
                active_experts = profile.moe.num_experts_per_tok + profile.moe.num_shared_experts
                total_experts = profile.moe.num_routed_experts + profile.moe.num_shared_experts
                if total_experts > 0:
                    moe_active_ratio = active_experts / total_experts
            decode_est = estimate_decode(
                profile=profile,
                total_weight_bytes=weight.total_bytes.value,
                gpu=gpu_spec,
                num_gpus=chosen,
                bw_utilization=decode_bw_utilization,
                moe_active_params_ratio=moe_active_ratio,
            )
            # Compute cluster headroom at the chosen tier + KV per request at the
            # *longest* surveyed context (most conservative).
            chosen_option = next(
                (o for o in fleet.options if o.gpu_count == chosen),
                fleet.options[-1],
            )
            headroom_per_gpu = (
                chosen_option.usable_bytes_per_gpu - chosen_option.weight_bytes_per_gpu
            )
            # Cluster-wide headroom is per-GPU * N; currently we use per-GPU view below.
            # Reference context for the L bound: match K's headroom context (128K
            # if model supports it, else max).
            kv_ref_ctx = 131_072 if 131_072 in kv_by_ctx else max(kv_by_ctx.keys())
            kv_ref_bytes: int = kv_by_ctx[kv_ref_ctx].value
            # Apply TP-aware sharding (same rule fleet planner uses).
            from llm_cal.fleet.planner import _kv_shards

            shards = _kv_shards(profile, chosen)
            kv_ref_per_gpu = max(1, kv_ref_bytes // shards)
            # Request KV lives per-GPU; under replication, it's the same value on all.
            # We compare cluster headroom against per-GPU KV (each request consumes
            # per-GPU KV on every rank simultaneously).
            # To convert to "how many requests fit", we divide *per-GPU* headroom
            # by *per-GPU* KV.
            headroom_per_req_view = max(0, headroom_per_gpu)
            concurrency_est = analyze_concurrency(
                cluster_headroom_bytes=headroom_per_req_view,
                kv_bytes_per_request=kv_ref_per_gpu,
                decode=decode_est,
                target_tokens_per_sec=eff_target,
                degradation=concurrency_degradation,
            )

        return EvaluationReport(
            model_id=model_id,
            source=artifact.source,
            commit_sha=artifact.commit_sha,
            gpu=gpu,
            gpu_spec=gpu_spec,
            gpu_error=gpu_error,
            engine=engine,
            profile=profile,
            weight=weight,
            total_params_estimate=total_params_est,
            reconciliation=reconciliation,
            kv_cache_by_context=kv_by_ctx,
            engine_match=engine_match,
            fleet=fleet,
            generated_command=generated_command,
            prefill=prefill_est,
            decode=decode_est,
            concurrency=concurrency_est,
            perf_input_tokens=input_tokens or 2000 if fleet else None,
            perf_output_tokens=output_tokens or 512 if fleet else None,
            perf_target_tokens_per_sec=target_tokens_per_sec or 30.0 if fleet else None,
        )

    def _fetch(self, model_id: str, refresh: bool) -> ModelArtifact:
        artifact = self._source.fetch(model_id)
        key = CacheKey(
            source=self._source.name,
            model_id=model_id,
            commit_sha=artifact.commit_sha,
        )
        cached = self._cache.get(key, bypass=refresh)
        if cached is not None:
            return cached
        self._cache.set(key, artifact)
        return artifact

    def _resolve_quant_fingerprint(
        self,
        artifact: ModelArtifact,
        observed_bytes: int,
        total_params: int,
    ) -> QuantFingerprint | None:
        """Resolve the quantization scheme via authoritative evidence.

        Priority:
          1. config.json `quantization_config` — explicit author declaration.
             Free, no extra network call. But if its predicted bytes are
             wildly off (>15% from observed), fall through — config.json
             can be incomplete or stale (DeepSeek-V4-Flash declares
             `quant_method=fp8` but ships an FP4+FP8 mixed pack; trusting
             the declaration produces a 45% wrong answer).
          2. safetensors file header — per-tensor dtype fingerprint. One
             Range GET on the first shard. Ground truth.

        Returns None on any failure. The reconciler falls back to bytes-only
        argmin in that case (v0.1.1 behavior).
        """
        fp = from_config(artifact.config)
        if fp is not None and self._fingerprint_matches_bytes(fp, observed_bytes, total_params):
            return fp

        shard = pick_sample_shard(artifact.siblings)
        if shard is None:
            return fp  # safetensors unavailable — best we can do is the config hint

        dtypes = fetch_tensor_dtypes(
            source=artifact.source,
            model_id=artifact.model_id,
            revision=artifact.commit_sha or "main",
            shard_filename=shard.filename,
        )
        if not dtypes:
            return fp

        st_fp = from_safetensors_dtypes(dtypes)
        # Header is ground truth — prefer it over config when both exist.
        return st_fp if st_fp is not None else fp

    @staticmethod
    def _fingerprint_matches_bytes(
        fp: QuantFingerprint, observed_bytes: int, total_params: int
    ) -> bool:
        """Sanity-check a fingerprint's predicted bytes against observed.

        Returns True if the declared scheme's predicted bytes are within 15%
        of observed. False means config.json is either lying or describes
        only part of the model — we should consult safetensors instead.
        """
        from llm_cal.weight_analyzer import _QUANT_BPP

        bpp = _QUANT_BPP.get(fp.scheme, 0.0)
        if bpp <= 0 or total_params <= 0 or observed_bytes <= 0:
            return True  # can't verify — don't penalize the fingerprint
        predicted = bpp * total_params
        rel_err = abs(observed_bytes - predicted) / predicted
        return rel_err <= 0.15

    @staticmethod
    def _select_context_lengths(profile: ArchitectureProfile, override: int | None) -> list[int]:
        if override is not None:
            return [override]
        candidates = [4_096, 32_768, 131_072]
        max_pos = profile.position.max_position_embeddings if profile.position else None
        if max_pos and max_pos > 131_072:
            candidates.append(max_pos)
        if max_pos:
            candidates = [c for c in candidates if c <= max_pos]
        return candidates


def _generate_command(
    engine: str,
    model_id: str,
    profile: ArchitectureProfile,
    tp: int,
    entry: EngineCompatEntry | None,
    max_model_len: int | None,
) -> str:
    engine_norm = engine.lower().strip()
    if engine_norm == "sglang":
        return generate_sglang_command(model_id, profile, tp, entry, max_model_len=max_model_len)
    return generate_vllm_command(model_id, profile, tp, entry, max_model_len=max_model_len)

File: crates/llm-infer-cal-core/src/core/evaluator.rs
use std::collections::BTreeMap;

use crate::architecture::detector::detect;
use crate::architecture::formulas::kv_cache::compute_kv_cache_bytes;
use crate::architecture::formulas::weight::estimate_total_params;
use crate::architecture::profile::ArchitectureProfile;
use crate::command_generator::sglang::generate_sglang_command;
use crate::command_generator::vllm::generate_vllm_command;
use crate::core::cache::{ArtifactCache, CacheKey};
use crate::engine_compat::loader::find_match;
use crate::engine_compat::EngineCompatEntry;
use crate::fleet::planner::{kv_shards, plan, FleetRecommendation};
use crate::hardware::loader::{lookup, GPUSpec};
use crate::model_source::base::{ModelArtifact, ModelSource, ModelSourceError};
use crate::model_source::huggingface::HuggingFaceSource;
use crate::output::labels::AnnotatedValue;
use crate::performance::compute::{
    estimate_decode, estimate_prefill, DecodeEstimate, PrefillEstimate,
    DEFAULT_CLUSTER_COMM_EFFICIENCY, DEFAULT_DECODE_BW_UTILIZATION, DEFAULT_PREFILL_UTILIZATION,
};
use crate::performance::concurrency::{analyze as analyze_concurrency, ConcurrencyAnalysis};
use crate::weight_analyzer::fingerprint::{from_config, from_safetensors_dtypes, QuantFingerprint};
use crate::weight_analyzer::reconciler::{reconcile, ReconciliationReport};
use crate::weight_analyzer::safetensors_reader::{fetch_tensor_dtypes, pick_sample_shard};
use crate::weight_analyzer::{analyze, WeightReport};

const KV_REFERENCE_CTX: u64 = 131_072;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct EvaluationOptions {
    pub gpu_count: Option<u64>,
    pub context_length: Option<u64>,
    pub refresh: bool,
    pub input_tokens: Option<u64>,
    pub output_tokens: Option<u64>,
    pub target_tokens_per_sec: Option<f64>,
    pub prefill_utilization: f64,
    pub decode_bw_utilization: f64,
    pub concurrency_degradation: f64,
}

impl Default for EvaluationOptions {
    fn default() -> Self {
        Self {
            gpu_count: None,
            context_length: None,
            refresh: false,
            input_tokens: None,
            output_tokens: None,
            target_tokens_per_sec: None,
            prefill_utilization: DEFAULT_PREFILL_UTILIZATION,
            decode_bw_utilization: DEFAULT_DECODE_BW_UTILIZATION,
            concurrency_degradation: 1.0,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct EvaluationReport {
    pub model_id: String,
    pub source: String,
    pub commit_sha: Option<String>,
    pub gpu: String,
    pub gpu_spec: Option<GPUSpec>,
    pub gpu_error: Option<String>,
    pub engine: String,
    pub profile: ArchitectureProfile,
    pub weight: WeightReport,
    pub total_params_estimate: AnnotatedValue<u64>,
    pub reconciliation: ReconciliationReport,
    pub kv_cache_by_context: BTreeMap<u64, AnnotatedValue<u64>>,
    pub engine_match: Option<EngineCompatEntry>,
    pub fleet: Option<FleetRecommendation>,
    pub generated_command: Option<String>,
    pub prefill: Option<PrefillEstimate>,
    pub decode: Option<DecodeEstimate>,
    pub concurrency: Option<ConcurrencyAnalysis>,
    pub perf_input_tokens: Option<u64>,
    pub perf_output_tokens: Option<u64>,
    pub perf_target_tokens_per_sec: Option<f64>,
}

pub struct Evaluator {
    source: Box<dyn ModelSource>,
    cache: Option<ArtifactCache>,
}

impl Default for Evaluator {
    fn default() -> Self {
        Self::new(
            Box::<HuggingFaceSource>::default(),
            ArtifactCache::with_default_ttl(None).ok(),
        )
    }
}

impl Evaluator {
    pub fn new(source: Box<dyn ModelSource>, cache: Option<ArtifactCache>) -> Self {
        Self { source, cache }
    }

    pub fn without_cache(source: Box<dyn ModelSource>) -> Self {
        Self::new(source, None)
    }

    pub fn evaluate(
        &self,
        model_id: &str,
        gpu: &str,
        engine: &str,
        options: EvaluationOptions,
    ) -> Result<EvaluationReport, ModelSourceError> {
        let artifact = self.fetch(model_id, options.refresh)?;
        let profile = detect(&artifact.config);

        let total_params_estimate = estimate_total_params(&profile);
        let total_params = total_params_estimate.value;

        let observed_bytes_for_fp = artifact
            .siblings
            .iter()
            .filter(|sibling| sibling.filename.ends_with(".safetensors"))
            .map(|sibling| sibling.size.unwrap_or(0))
            .sum::<u64>();
        let fingerprint = self.resolve_quant_fingerprint(
            &artifact,
            observed_bytes_for_fp,
            if total_params > 0 { total_params } else { 0 },
        );
        let weight = analyze(
            &artifact.siblings,
            (total_params > 0).then_some(total_params),
            fingerprint.as_ref(),
        );
        let reconciliation = reconcile(
            weight.total_bytes.value,
            if total_params > 0 { total_params } else { 1 },
            fingerprint.as_ref(),
        );

        let contexts_to_report = select_context_lengths(&profile, options.context_length);
        let kv_cache_by_context = contexts_to_report
            .into_iter()
            .map(|ctx| (ctx, compute_kv_cache_bytes(&profile, ctx, 2)))
            .collect::<BTreeMap<_, _>>();

        let engine_match = find_match(engine, &profile.model_type, None, None);

        let (gpu_spec, gpu_error) = match lookup(gpu) {
            Ok(spec) => (Some(spec), None),
            Err(error) => (None, Some(error.to_string())),
        };

        let mut fleet = None;
        let mut generated_command = None;
        if let Some(spec) = &gpu_spec {
            if weight.total_bytes.value > 0 {
                let kv_ref = compute_kv_cache_bytes(&profile, KV_REFERENCE_CTX, 2);
                let kv_by_context_bytes = kv_cache_by_context
                    .iter()
                    .filter_map(|(ctx, value)| (value.value > 0).then_some((*ctx, value.value)))
                    .collect::<Vec<_>>();
                let recommendation = plan(
                    &profile,
                    weight.total_bytes.value,
                    kv_ref.value.max(1),
                    spec,
                    options.gpu_count,
                    &kv_by_context_bytes,
                );
                let chosen_count = options.gpu_count.unwrap_or_else(|| {
                    recommendation
                        .options
                        .iter()
                        .find(|option| option.tier == recommendation.best_tier)
                        .or_else(|| recommendation.options.first())
                        .map(|option| option.gpu_count)
                        .unwrap_or(1)
                });
                generated_command = Some(generate_command(
                    engine,
                    model_id,
                    &profile,
                    chosen_count,
                    engine_match.as_ref(),
                    options.context_length,
                ));
                fleet = Some(recommendation);
            }
        }

        let mut prefill = None;
        let mut decode = None;
        let mut concurrency = None;
        if let (Some(spec), Some(recommendation)) = (&gpu_spec, &fleet) {
            if total_params > 0 {
                let chosen = options.gpu_count.unwrap_or_else(|| {
                    recommendation
                        .options
                        .iter()
                        .find(|option| option.tier == recommendation.best_tier)
                        .or_else(|| recommendation.options.first())
                        .map(|option| option.gpu_count)
                        .unwrap_or(1)
                });
                let input_tokens = options.input_tokens.unwrap_or(2000);
                let target_tokens_per_sec = options.target_tokens_per_sec.unwrap_or(30.0);

                prefill = Some(estimate_prefill(
                    &profile,
                    total_params,
                    spec,
                    chosen,
                    input_tokens,
                    options.prefill_utilization,
                ));

                let moe_active_ratio = profile.moe.as_ref().and_then(|moe| {
                    let active = moe.num_experts_per_tok + moe.num_shared_experts;
                    let total = moe.num_routed_experts + moe.num_shared_experts;
                    (total > 0).then_some(active as f64 / total as f64)
                });

                let decode_estimate = estimate_decode(
                    &profile,
                    weight.total_bytes.value,
                    spec,
                    chosen,
                    options.decode_bw_utilization,
                    DEFAULT_CLUSTER_COMM_EFFICIENCY,
                    moe_active_ratio,
                );

                let chosen_option = recommendation
                    .options
                    .iter()
                    .find(|option| option.gpu_count == chosen)
                    .or_else(|| recommendation.options.last());
                if let Some(chosen_option) = chosen_option {
                    if let Some((kv_ref_ctx, kv_ref_bytes)) =
                        reference_context_bytes(&kv_cache_by_context)
                    {
                        let _ = kv_ref_ctx;
                        let headroom_per_gpu = chosen_option
                            .usable_bytes_per_gpu
                            .saturating_sub(chosen_option.weight_bytes_per_gpu);
                        let shards = kv_shards(&profile, chosen).max(1);
                        let kv_ref_per_gpu = (kv_ref_bytes / shards).max(1);
                        concurrency = Some(analyze_concurrency(
                            headroom_per_gpu,
                            kv_ref_per_gpu,
                            &decode_estimate,
                            target_tokens_per_sec,
                            options.concurrency_degradation,
                        ));
                    }
                }
                decode = Some(decode_estimate);
            }
        }

        let has_fleet = fleet.is_some();
        Ok(EvaluationReport {
            model_id: model_id.to_string(),
            source: artifact.source,
            commit_sha: artifact.commit_sha,
            gpu: gpu.to_string(),
            gpu_spec,
            gpu_error,
            engine: engine.to_string(),
            profile,
            weight,
            total_params_estimate,
            reconciliation,
            kv_cache_by_context,
            engine_match,
            fleet,
            generated_command,
            prefill,
            decode,
            concurrency,
            perf_input_tokens: has_fleet.then_some(options.input_tokens.unwrap_or(2000)),
            perf_output_tokens: has_fleet.then_some(options.output_tokens.unwrap_or(512)),
            perf_target_tokens_per_sec: has_fleet
                .then_some(options.target_tokens_per_sec.unwrap_or(30.0)),
        })
    }

    fn fetch(&self, model_id: &str, refresh: bool) -> Result<ModelArtifact, ModelSourceError> {
        let artifact = self.source.fetch(model_id)?;
        let key = CacheKey::new(self.source.name(), model_id, artifact.commit_sha.as_deref());
        if let Some(cache) = &self.cache {
            if let Ok(Some(cached)) = cache.get(&key, refresh) {
                return Ok(cached);
            }
            let _ = cache.set(&key, &artifact);
        }
        Ok(artifact)
    }

    fn resolve_quant_fingerprint(
        &self,
        artifact: &ModelArtifact,
        observed_bytes: u64,
        total_params: u64,
    ) -> Option<QuantFingerprint> {
        let fp = from_config(&artifact.config);
        if fp
            .as_ref()
            .is_some_and(|fp| fingerprint_matches_bytes(fp, observed_bytes, total_params))
        {
            return fp;
        }

        let shard = match pick_sample_shard(&artifact.siblings) {
            Some(shard) => shard,
            None => return fp,
        };
        let revision = artifact
            .commit_sha
            .as_deref()
            .unwrap_or(match artifact.source.as_str() {
                "modelscope" => "master",
                _ => "main",
            });
        let dtypes = match fetch_tensor_dtypes(
            &artifact.source,
            &artifact.model_id,
            revision,
            &shard.filename,
            None,
        ) {
            Some(dtypes) => dtypes,
            None => return fp,
        };

        from_safetensors_dtypes(&dtypes).or(fp)
    }
}

fn fingerprint_matches_bytes(
    fp: &QuantFingerprint,
    observed_bytes: u64,
    total_params: u64,
) -> bool {
    let bpp = fp.scheme.bpp();
    if bpp <= 0.0 || total_params == 0 || observed_bytes == 0 {
        return true;
    }
    let predicted = bpp * total_params as f64;
    let rel_err = (observed_bytes as f64 - predicted).abs() / predicted;
    rel_err <= 0.15
}

fn select_context_lengths(profile: &ArchitectureProfile, override_ctx: Option<u64>) -> Vec<u64> {
    if let Some(override_ctx) = override_ctx {
        return vec![override_ctx];
    }

    let mut candidates = vec![4_096, 32_768, KV_REFERENCE_CTX];
    let max_pos = profile
        .position
        .as_ref()
        .and_then(|position| position.max_position_embeddings);
    if let Some(max_pos) = max_pos {
        if max_pos > KV_REFERENCE_CTX {
            candidates.push(max_pos);
        }
        candidates.retain(|ctx| *ctx <= max_pos);
    }
    candidates
}

fn reference_context_bytes(
    kv_cache_by_context: &BTreeMap<u64, AnnotatedValue<u64>>,
) -> Option<(u64, u64)> {
    if let Some(value) = kv_cache_by_context.get(&KV_REFERENCE_CTX) {
        return Some((KV_REFERENCE_CTX, value.value));
    }
    kv_cache_by_context
        .iter()
        .next_back()
        .map(|(ctx, value)| (*ctx, value.value))
}

fn generate_command(
    engine: &str,
    model_id: &str,
    profile: &ArchitectureProfile,
    tp: u64,
    entry: Option<&EngineCompatEntry>,
    max_model_len: Option<u64>,
) -> String {
    if engine.trim().eq_ignore_ascii_case("sglang") {
        generate_sglang_command(model_id, profile, tp, entry, max_model_len)
    } else {
        generate_vllm_command(model_id, profile, tp, entry, max_model_len)
    }
}
